sds/src/secondary_validation/nir_checksum.rs

```diff
@@ -0,0 +1,75 @@
+use std::str::Chars;
+
+use crate::secondary_validation::Validator;
+
+pub struct NirChecksum;
+
+impl Validator for NirChecksum {
+    fn is_valid_match(&self, regex_match: &str) -> bool {
+        let mut input_iter = regex_match.chars();
+
+        let digit: Option<i64> = get_next_digit_chars(&mut input_iter, 13)
+            .and_then(|chars| chars.into_iter().collect::<String>().parse().ok());
+        if digit.is_none() {
+            return false;
+        }
+
+        let checksum: Option<u8> = get_next_digit_chars(&mut input_iter, 2)
+            .and_then(|chars| chars.into_iter().collect::<String>().parse().ok());
+        if checksum.is_none() || get_next_digit_str(&mut input_iter).is_some() {
+            return false;
+        }
+
+        (97 - digit.unwrap() % 97) as u8 == checksum.unwrap()
+    }
+}
+
+fn get_next_digit_str(chars: &mut Chars<'_>) -> Option<char> {
+    chars.find(|&char| char.is_ascii_digit())
+}
+
+fn get_next_digit_chars(chars: &mut Chars<'_>, size: usize) -> Option<Vec<char>> {
```

Comment:
This function seems to be unnecessarily allocating. Chars are collected and put in a `Vec`, which is immediately iterated over again and collected into a `String`, which is then immediately used to converted into a number. Ideally this would directly compute the integer representation do no allocation is needed at all.

If that is too much to do right now, the chars could at least be collected immediately into a `String` and that is returned instead of the intermediate `Vec`.

Reply:
I was not sure of how to do this, this push me to find a workaround in https://github.com/DataDog/dd-sensitive-data-scanner/pull/88/commits/2125da14fcf39f78732226636b22bf197175b845